crates/compiler/src/compiler.rs

```diff
@@ -1,18 +1,237 @@
+
+        // ok now we will add in our lastline tags
+        // we do this BEFORE we build our strings table otherwise the tags will get missed
+        // this should probably be a flag instead of every time though
+        // TODO
+        // let lastLineTagger = new LastLineBeforeOptionsVisitor();
+        //lastLineTagger.Visit(parseResult.Tree);
+        let mut visitor =
+            StringTableGeneratorVisitor::new(file.file_name.clone(), string_table_manager.clone());
+        visitor.visit(&*parse_result.tree);
+        previous.diagnostics.extend(visitor.diagnostics);
+        string_table_manager.extend(visitor.string_table_manager);
+        parsed_files.push(parse_result);
+    }
+    previous.string_table = string_table_manager.into();
+    previous
+}
+
+fn parse_syntax_tree<'a>(file: &'a File, diagnostics: &mut Vec<Diagnostic>) -> FileParseResult<'a> {
+    let input = CodePoint8BitCharStream::new(file.source.as_bytes());
+    let mut lexer = YarnSpinnerLexer::new(input);
+
+    // turning off the normal error listener and using ours
+    let file_name = file.file_name.clone();
+    let lexer_error_listener = LexerErrorListener::new(file_name.clone());
+    let lexer_error_listener_diagnostics = lexer_error_listener.diagnostics.clone();
+    lexer.remove_error_listeners();
+    lexer.add_error_listener(Box::new(lexer_error_listener));
+
+    let tokens = CommonTokenStream::new(lexer);
+    let mut parser = YarnSpinnerParser::with_strategy(tokens, ErrorStrategy::new());
+    let parser_error_listener = ParserErrorListener::new(file.clone());
+    let parser_error_listener_diagnostics = parser_error_listener.diagnostics.clone();
+
+    parser.remove_error_listeners();
+    parser.add_error_listener(Box::new(parser_error_listener));
+
+    let lexer_error_listener_diagnostics_borrowed = lexer_error_listener_diagnostics.borrow();
+    let parser_error_listener_diagnostics_borrowed = parser_error_listener_diagnostics.borrow();
+    let new_diagnostics = lexer_error_listener_diagnostics_borrowed
+        .iter()
+        .chain(parser_error_listener_diagnostics_borrowed.iter())
+        .cloned();
+    diagnostics.extend(new_diagnostics);
+
+    FileParseResult::new(file_name, parser)
+}
+
+pub(crate) fn get_line_id_for_node_name(name: &str) -> String {
+    format!("line:{name}")
+}
+
+pub(crate) fn add_hashtag_child<'input>(
+    parent: &impl YarnSpinnerParserContext<'input>,
+    text: String,
+) {
+    // Hack: need to convert the reference to an Rc somehow
+    let parent = parent.get_children().next().unwrap().get_parent().unwrap();
+    // Taken from C# implementation of `CommonToken`s constructor
+    let string_id_token = CommonTokenFactory.create::<InputStream<&'input str>>(
+        None,
+        HASHTAG_TEXT,
+        text.into(),
+        0,
+        0,
+        0,
+        0,
+        -1,
+    );
+    // `new_with_text` was hacked into the generated parser. Also, `FooContextExt::new` is usually private...
```

Comment:
is that documented in the readme?

Reply:
Yeah, it's part of the PR already :P
https://github.com/Mafii/rusty-yarn-spinner/pull/17/files#diff-33b29f37c67b2f1449782135a34962699adc69f74475770e11e02faca24e67a5

Reply:
Is the documentation okay like that?